File: frontend/js/views/seller_financing.js
// Seller financing / carryback note — amortized payment, balloon balance,
// and the seller's interest income, via /calc/seller-financing. Live.

import { api } from '../api.js';
import { applyUiI18n, t } from '../i18n.js';
import { currentViewToken, viewIsCurrent } from '../app.js';
import { showToast } from '../toast.js';
import { debounce } from '../util.js';
import * as enh from '../calc_enhance.js';

const FIELDS = [
    ['sale_price_usd', 'Sale price ($)', 400000],
    ['down_payment_usd', 'Down payment ($)', 80000],
    ['annual_rate_pct', 'Note rate (%)', 7],
    ['amortization_years', 'Amortization (years)', 30],
    ['balloon_years', 'Balloon (years, 0 = none)', 5],
];

const money = (n) => '$' + Number(n).toLocaleString(undefined, { maximumFractionDigits: 2 });
const money0 = (n) => '$' + Number(n).toLocaleString(undefined, { maximumFractionDigits: 0 });
const VIEW = 'seller-financing';
let lastReport = null;
let lastBody = null;

export async function renderSellerFinancing(mount, _appState) {
    const tok = currentViewToken();
    if (!viewIsCurrent(tok)) return;
    mount.innerHTML = `
        <h1 class="view-title"><span data-i18n="view.sf.h1.title">// SELLER FINANCING</span></h1>
        <p class="muted small" data-i18n="view.sf.hint.intro">
            When the seller is the bank: the buyer puts down a down payment and pays the seller
            directly on the balance (the carryback note) at a rate, amortized over a schedule —
            often with a balloon where the remaining balance is due in a few years. Shows the
            monthly payment, the balloon balance owed at that date, and the interest the seller
            earns. Updates as you type.
        </p>
        <div class="lpv-split">
        <div class="chart-panel">
            <h2 data-i18n="view.sf.h2.inputs">The deal</h2>
            <form id="sf-form" class="inline-form">
                ${FIELDS.map(([key, label, def]) => `
                    <label><span data-i18n="view.sf.label.${key}">${label}</span>
                        <input type="number" step="0.01" min="0" name="${key}" value="${def}" required></label>
                `).join('')}
            </form>
            <div id="sf-tools" class="ce-toolbar"></div>
            <button type="button" id="sf-sens-btn" class="ce-tool" data-i18n="calc.enh.sens.run">▦ Sensitivity</button>
            <div id="sf-sens" class="ce-sens"></div>
        </div>
        <div id="sf-result" class="lpv-preview"></div>
        </div>
    `;
    applyUiI18n(mount);

    const form = mount.querySelector('#sf-form');
    enh.prefillForm(form, enh.readHashInputs());
    const readBody = () => {
        const fd = new FormData(form);
        const body = {};
        for (const [key] of FIELDS) body[key] = Number(fd.get(key)) || 0;
        return body;
    };
    const generate = async () => {
        const body = readBody();
        try {
            const r = await api.calcSellerFinancing(body);
            if (!viewIsCurrent(tok)) return;
            lastReport = r; lastBody = body;
            renderResult(mount, r, body, tok);
        } catch (err) {
            showToast(err.message || t('view.sf.toast.error'), { level: 'error' });
        }
    };
    enh.mountToolbar(mount.querySelector('#sf-tools'), { viewId: VIEW, getInputs: () => lastBody || readBody(), getRows: () => reportRows(lastReport), filename: 'seller-financing.csv' });
    mount.querySelector('#sf-sens-btn').addEventListener('click', () => runSens(mount, readBody(), tok));
    form.addEventListener('input', debounce(generate, 250));
    form.addEventListener('submit', (e) => { e.preventDefault(); generate(); });
    generate();
}

function reportRows(r) {
    if (!r) return [];
    return [
        ['metric', 'value'],
        ['monthly_payment_usd', r.monthly_payment_usd],
        ['note_amount_usd', r.note_amount_usd],
        ['balloon_balance_usd', r.balloon_balance_usd],
        ['seller_interest_income_usd', r.seller_interest_income_usd],
        ['total_payments_usd', r.total_payments_usd],
    ];
}

async function renderResult(mount, r, body, tok) {
    const el = mount.querySelector('#sf-result');
    // Line chart: monthly payment as the note rate sweeps 0 -> 12%.
    const xs = enh.linspace(0, 12, 13);
    const pts = await Promise.all(xs.map(async (rate) => {
        const rr = await api.calcSellerFinancing({ ...body, annual_rate_pct: rate });
        return { x: rate, y: rr ? rr.monthly_payment_usd : NaN };
    }));
    if (!viewIsCurrent(tok)) return;
    const chart = enh.svgLineChart(pts, { xlabel: 'rate %', ylabel: 'payment $/mo' });
    el.innerHTML = `
        <div class="chart-panel">
            <h2 data-i18n="view.sf.h2.result">The note</h2>
            <div class="cards">
                <div class="card pos"><div class="label" data-i18n="view.sf.card.payment">Monthly payment</div>
                    <div class="value pos">${money(r.monthly_payment_usd)}</div></div>
                <div class="card"><div class="label" data-i18n="view.sf.card.note">Note amount</div>
                    <div class="value">${money0(r.note_amount_usd)}</div></div>
                <div class="card ${r.has_balloon ? 'neg' : ''}"><div class="label" data-i18n="view.sf.card.balloon">Balloon balance</div>
                    <div class="value">${r.has_balloon ? money0(r.balloon_balance_usd) : t('view.sf.none')}</div></div>
                <div class="card"><div class="label" data-i18n="view.sf.card.interest">Seller interest income</div>
                    <div class="value">${money0(r.seller_interest_income_usd)}</div></div>
            </div>
            ${chart}
            <table class="data-table">
                <thead><tr><th data-i18n="view.sf.col.line">Line</th><th data-i18n="view.sf.col.amount">Amount</th></tr></thead>
                <tbody>
                    <tr><td>${t('view.sf.row.payments')} ${r.has_balloon ? t('view.sf.to_balloon') : t('view.sf.full_term')}</td><td>${money0(r.total_payments_usd)}</td></tr>
                    <tr><td data-i18n="view.sf.row.balloon">Balloon balance due</td><td>${money0(r.balloon_balance_usd)}</td></tr>
                    <tr class="emph"><td data-i18n="view.sf.row.interest">Seller interest income</td><td class="pos">${money0(r.seller_interest_income_usd)}</td></tr>
                </tbody>
            </table>
        </div>
    `;
    applyUiI18n(el);
}

async function runSens(mount, base, tok) {
    const panel = mount.querySelector('#sf-sens');
    panel.innerHTML = `<p class="muted small" data-i18n="calc.enh.sens.running">Computing…</p>`; applyUiI18n(panel);
    // x: note rate 0 -> 12%; y: amortization 10 -> 40 years. Output: monthly payment.
    const xVals = enh.linspace(0, 12, 5);
    const yVals = enh.linspace(10, 40, 5);
    const { cells } = await enh.runSensitivity({ base, xKey: 'annual_rate_pct', yKey: 'amortization_years', xVals, yVals: yVals.map(Math.round), compute: (b) => api.calcSellerFinancing(b), pick: (r) => (r ? r.monthly_payment_usd : null) });
    if (!viewIsCurrent(tok)) return;
    panel.innerHTML = enh.renderSensitivityTable({ xVals, yVals: yVals.map(Math.round), cells, fmt: (v) => (v == null ? '—' : '$' + Math.round(v)), xfmt: (v) => v.toFixed(1) + '%', yfmt: (v) => v.toFixed(0) + 'y', xName: t('view.sf.label.annual_rate_pct') || 'Rate', yName: t('view.sf.label.amortization_years') || 'Amort' });
}

File: crates/traderview-core/src/seller_financing.rs
//! Seller financing / carryback note — when the seller is the bank.
//!
//! Instead of the buyer getting a bank loan, the seller "carries back" a
//! note: the buyer puts down a down payment and pays the seller directly on
//! the balance, at a rate, amortized over a schedule — often with a
//! **balloon** where the remaining balance is due in a few years.
//!
//!   * note amount = sale price − down payment
//!   * monthly payment amortizes the note over the amortization period
//!   * balloon balance = the loan balance still owed at the balloon date
//!   * seller interest income = total payments + balloon − note amount
//!
//! Pure compute (standard amortization, zero-rate straight-line guard).

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Deserialize)]
pub struct SellerFinancingInput {
    pub sale_price_usd: f64,
    pub down_payment_usd: f64,
    pub annual_rate_pct: f64,
    /// Years the payment is amortized over (e.g. 30).
    pub amortization_years: f64,
    /// Years until the balloon (remaining balance due); 0 = fully amortizing.
    #[serde(default)]
    pub balloon_years: f64,
}

#[derive(Debug, Clone, Serialize)]
pub struct SellerFinancingResult {
    pub note_amount_usd: f64,
    pub monthly_payment_usd: f64,
    /// Balance still owed at the balloon date (0 if fully amortizing).
    pub balloon_balance_usd: f64,
    pub has_balloon: bool,
    /// Payments made before the balloon (or over the full term).
    pub total_payments_usd: f64,
    /// Interest the seller earns over the life of the note.
    pub seller_interest_income_usd: f64,
}

fn monthly_payment(principal: f64, annual_rate_pct: f64, months: f64) -> f64 {
    if months <= 0.0 {
        return 0.0;
    }
    let r = annual_rate_pct / 100.0 / 12.0;
    if r.abs() < 1e-12 {
        principal / months
    } else {
        let f = (1.0 + r).powf(months);
        principal * r * f / (f - 1.0)
    }
}

/// Remaining balance after `k` payments of `pmt` on `principal` at monthly `r`.
fn balance_after(principal: f64, annual_rate_pct: f64, pmt: f64, k: f64) -> f64 {
    let r = annual_rate_pct / 100.0 / 12.0;
    if r.abs() < 1e-12 {
        (principal - pmt * k).max(0.0)
    } else {
        let f = (1.0 + r).powf(k);
        (principal * f - pmt * (f - 1.0) / r).max(0.0)
    }
}

pub fn analyze(i: &SellerFinancingInput) -> SellerFinancingResult {
    let note = (i.sale_price_usd - i.down_payment_usd).max(0.0);
    let amort_months = (i.amortization_years * 12.0).max(0.0);
    let pmt = monthly_payment(note, i.annual_rate_pct, amort_months);

    // The note runs to the balloon, or to full amortization if no balloon.
    let has_balloon = i.balloon_years > 0.0 && i.balloon_years < i.amortization_years;
    let pay_months = if has_balloon { i.balloon_years * 12.0 } else { amort_months };

    let balloon_balance = if has_balloon {
        balance_after(note, i.annual_rate_pct, pmt, pay_months)
    } else {
        0.0
    };

    let total_payments = pmt * pay_months;
    let seller_interest = total_payments + balloon_balance - note;

    SellerFinancingResult {
        note_amount_usd: note,
        monthly_payment_usd: pmt,
        balloon_balance_usd: balloon_balance,
        has_balloon,
        total_payments_usd: total_payments,
        seller_interest_income_usd: seller_interest,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn base() -> SellerFinancingInput {
        SellerFinancingInput {
            sale_price_usd: 400_000.0,
            down_payment_usd: 80_000.0,
            annual_rate_pct: 7.0,
            amortization_years: 30.0,
            balloon_years: 5.0,
        }
    }

    #[test]
    fn note_is_price_minus_down() {
        let r = analyze(&base());
        assert!((r.note_amount_usd - 320_000.0).abs() < 1e-6);
    }

    #[test]
    fn monthly_payment_amortizes_note() {
        // 320k @ 7% / 30y ≈ $2,128.97.
        let r = analyze(&base());
        assert!((r.monthly_payment_usd - 2_128.97).abs() < 0.5);
    }

    #[test]
    fn balloon_balance_after_five_years() {
        // Most of a 30y note is still owed after 5y.
        let r = analyze(&base());
        assert!(r.has_balloon);
        assert!(r.balloon_balance_usd > 280_000.0 && r.balloon_balance_usd < 310_000.0);
    }

    #[test]
    fn no_balloon_fully_amortizes() {
        let r = analyze(&SellerFinancingInput { balloon_years: 0.0, ..base() });
        assert!(!r.has_balloon);
        assert!(r.balloon_balance_usd.abs() < 1e-6);
    }

    #[test]
    fn seller_interest_income_positive() {
        let r = analyze(&base());
        assert!(r.seller_interest_income_usd > 0.0);
        // total payments + balloon − note = interest.
        let expected = r.total_payments_usd + r.balloon_balance_usd - r.note_amount_usd;
        assert!((r.seller_interest_income_usd - expected).abs() < 1e-6);
    }

    #[test]
    fn higher_rate_raises_payment() {
        let low = analyze(&SellerFinancingInput { annual_rate_pct: 5.0, ..base() });
        let high = analyze(&SellerFinancingInput { annual_rate_pct: 9.0, ..base() });
        assert!(high.monthly_payment_usd > low.monthly_payment_usd);
    }

    #[test]
    fn balloon_at_or_after_amort_is_full_amortization() {
        let r = analyze(&SellerFinancingInput { balloon_years: 30.0, ..base() });
        assert!(!r.has_balloon); // not < amortization
        assert!(r.balloon_balance_usd.abs() < 1e-6);
    }

    #[test]
    fn zero_rate_straight_line_balance() {
        // 320k over 30y = 888.89/mo; after 5y (60 pmts) balance = 320k − 53,333 = 266,667.
        let r = analyze(&SellerFinancingInput { annual_rate_pct: 0.0, ..base() });
        assert!((r.monthly_payment_usd - 320_000.0 / 360.0).abs() < 1e-6);
        let expected = 320_000.0 - (320_000.0 / 360.0) * 60.0;
        assert!((r.balloon_balance_usd - expected).abs() < 1e-3);
    }
}
